design/sync-labels-bmh-to-node.md

```diff
@@ -0,0 +1,353 @@
+
+Scenario (i) can be handled as follows:
+
+1. CAPM3 will set up a watch on the BMH resource. It only requires read-only
+  access similar to what is done today:
+
+    baremetalhosts,verbs=get;list;watch
+
+2. On each BMH reconcile event, the controller will fetch the corresponding
+  Node. The _mapping_ can be accomplished by:
+
+    BMH.ConsumerRef --> Metal3Machine.OwnerRef --> Machine.NodeRef
+
+3. Synchronize the labels and update the Node in the workload cluster. CAPM3
+  already fetches the Node objects from the workload clusters today using a
+  remote client. We would leverage that same capability here.
+
+An alternative means to accomplish the mapping is to use the ProviderID:
+
+    BMH.ConsumerRef --> Metal3Machine.Spec.ProviderID --> Node.Spec.ProviderID
+
+Unless we can set up filters, the approach would require us to fetch all Nodes
+and then scan through them to find the Node matching the ProviderID. This does
+not seem scalable for large clusters.
+Field-selectors could potentially be employed and would require further
+investigation. However,there is no obvious advantage over the earlier approach
+for doing this mapping.
+
+Scenario (ii) can be handled with a full-sync of all BMH and Node objects at a
+preconfigured synchronization interval:
+
+1. At a configurable interval (e.g. 15 seconds), fetch complete list of BMHs.
+2. For each BMH, fetch corresponding Node using the mapping discussed above.
+3. Synchronize the labels and update the Node in the workload cluster.
+
+An alternative would be to set up a watch on the Node resource types in every
+workload cluster. The benefit being that we can quickly synchronize any label
+changes on the Node object itself. However, this introduces complications when
+the management cluster != workload cluster and additionally, when there are
+many workload clusters being managed by a single management cluster. How such
+a watch could be implemented within the controller-runtime model would require
+further investigation.
+
+### Configuration
+
+#### Command-Line Flag
+
+We need to expose two additional configuration variables to the user: the set
+of label prefixes and the full-sync interval. We can introduce two new
+command-line flags for CAPM3, and also utilize default values if the flags are
+not specified: (i) `--bmh-label-prefixes` with a nil default value and (ii)
+`--label-sync-interval` with a 15-second default value.
+
+Unfortunately, command-line flags will not be compatible with openShift where
+cluster administrators do not deploy metal3 themselves and do not have access
+to the command line arguments for the cluster-api provider.
+
+#### Annotations on BMH
+
+In this approach, an annotation is added to a BMH, in addition to the labels
+to be synchronized, which reflects the prefixes to be utilized. For example,
+
+```yaml
+  kind: BareMetalHost
+  metadata:
+    annotations:
+      cluster.x-k8s.io/label-prefixes: "my-prefix.metal3.io, my-prefix.kubernetes.io"
```

Comment:
Annotations can be non-string types, so we could use a list as the value.

Should the name for the annotation be scoped to metal3.io, or is this a cluster API feature?